### PR TITLE
RESP Charges

### PR DESCRIPTION
**Pull request by tkramer-motion**
Uses pyscf to compute RESP charges as an alternative to OpenEye's AM1BCC

### CODE REVIEW

File: timemachine/ff/handlers/nonbonded.py [hunk 1]
```diff
@@ -1,27 +1,44 @@
...
+from functools import partial
+from shutil import which
 
 import jax.numpy as jnp
 import networkx as nx
 import numpy as np
+from Auto3D.ASE.geometry import opt_geometry
 from jax import jit, vmap
 from numpy.typing import NDArray
+from openff.recharge.aromaticity import AromaticityModel
```
Comment (proteneer) [inline timemachine/ff/handlers/nonbonded.py:17]: is this overwritten by `from timemachine.ff.handlers.bcc_aromaticity import AromaticityModel` down below?

File: timemachine/ff/handlers/nonbonded.py [hunk 2]
```diff
@@ -1,27 +1,44 @@
...
+from openff.recharge.utilities.molecule import extract_conformers
+from openff.toolkit import unit, RDKitToolkitWrapper
```
Comment (proteneer) [inline timemachine/ff/handlers/nonbonded.py:19]: I think unit here is just `openmm.unit`

File: requirements.txt [hunk 3]
```diff
@@ -7,3 +7,9 @@ rdkit==2024.9.4
 --extra-index-url https://pypi.anaconda.org/OpenEye/simple
 openeye-toolkits==2020.2.0
 openmm==8.2.0
+gpu4pyscf-cuda12x
```
Comment (proteneer) [inline requirements.txt:10]: these should be versioned

File: timemachine/ff/handlers/nonbonded.py [hunk 4]
```diff
@@ -150,6 +169,297 @@ def oe_assign_charges(mol, charge_model=AM1BCCELF10):
...
+    rdmol = Chem.Mol(_rdmol)
+    rdkit_generate_conformations(rdmol)
+
+    print(f"Generated {rdmol.GetNumConformers()} RDKit conformers")
+
+    # Convert to OpenFF Molecule for ELF conformer selection
+    molecule: Molecule = Molecule.from_rdkit(rdmol)
+    # Apply ELF conformer selection to get diverse, representative conformers
+    # This helps ensure charges are computed from a representative ensemble
+    molecule.apply_elf_conformer_selection(
```
Comment (proteneer) [inline timemachine/ff/handlers/nonbonded.py:416]: ugh - I think this is the biggest dependency on the openff

File: requirements.txt [hunk 5]
```diff
@@ -7,3 +7,9 @@ rdkit==2024.9.4
...
+cutensor-cu12
+Auto3D
+torch
+torchvision
```
Comment (proteneer) [inline requirements.txt:14]: torchvision and torchaudio are needed?

File: timemachine/ff/handlers/nonbonded.py [hunk 6]
```diff
@@ -150,6 +169,297 @@ def oe_assign_charges(mol, charge_model=AM1BCCELF10):
...
+        # Attempt to optimize geometry using AIMNET neural network potential
+        # This can provide better initial coordinates for QM calculations
+        try:
+            out_path = opt_geometry(os.path.join(tmpdir, "molecule.sdf"), model_name="AIMNET")
+            m = Chem.MolFromMolFile(out_path, removeHs=False)
+            bad_bond = False
+            # Check for unreasonable bond lengths that indicate optimization failure
+            for bnd in m.GetBonds():
+                bond_length = GetBondLength(m.GetConformer(), bnd.GetBeginAtomIdx(), bnd.GetEndAtomIdx())
+                if bond_length > 3.0:
```
Comment (proteneer) [inline timemachine/ff/handlers/nonbonded.py:266]: Might be better to have the threshold be % based as opposed to a hard cutoff - i.e. if the bond lengths change +- 15% that would be quite significant.

File: timemachine/ff/handlers/nonbonded.py [hunk 7]
```diff
@@ -150,6 +169,297 @@ def oe_assign_charges(mol, charge_model=AM1BCCELF10):
...
+
+            if not bad_bond:
+                # Use optimized coordinates if geometry is reasonable
+                xyz = make_xyz(m)
+            else:
+                # Fall back to original coordinates if optimization failed
+                xyz = make_xyz(rdmol)
+        except Exception as e:
+            # If optimization fails entirely, use original coordinates
+            print(e)
```
Comment (proteneer) [inline timemachine/ff/handlers/nonbonded.py:280]: wouldn't it be better to just abort here?

File: timemachine/ff/handlers/nonbonded.py [hunk 8]
```diff
@@ -150,6 +169,297 @@ def oe_assign_charges(mol, charge_model=AM1BCCELF10):
...
+                    -1,
+                    1.85,  # Br
+                ]
+            )
+        )
+
+        # First fit ESP charges without restraints
+        # This provides initial charges that exactly reproduce the electrostatic potential
+        esp.esp_solve(mol, dm, rad=rad)
+        print("Fitted ESP charge")
```
Comment (proteneer) [inline timemachine/ff/handlers/nonbonded.py:368]: is this comment necessary?

File: timemachine/ff/handlers/nonbonded.py [hunk 9]
```diff
@@ -150,6 +169,297 @@ def oe_assign_charges(mol, charge_model=AM1BCCELF10):
...
+
+        # First stage RESP fitting with default restraints
+        # This applies weak restraints to prevent overfitting
+        esp.resp_solve(mol, dm, rad=rad)
+        sum_constraints = []
+
+        # Second stage RESP fitting with symmetry constraints
```
Comment (proteneer) [inline timemachine/ff/handlers/nonbonded.py:375]: neat!

File: timemachine/ff/handlers/nonbonded.py [hunk 10]
```diff
@@ -150,6 +169,297 @@ def oe_assign_charges(mol, charge_model=AM1BCCELF10):
...
+    # Calculate RESP charges for each selected conformer
+    charges = []
+    energies = []
+    for chs, energy in map(partial(resp_assign_partial_charges, _rdmol), [[c] for c in conformers]):
+        if chs is not None:
+            charges.append(chs)
+            energies.append(energy)
+
+    # Convert charges list to numpy array for averaging
+    am1_partial_charges = np.array(charges)
```
Comment (proteneer) [inline timemachine/ff/handlers/nonbonded.py:434]: should this still be named "am1"?

File: tests/test_handlers.py [hunk 11]
```diff
@@ -456,6 +456,34 @@ def test_am1bcc_parameterization():
...
+
+    assert len(charges) == mol.GetNumAtoms()
+    assert mol.HasProp(cache_key)
+
+    cached_charges = am1h.parameterize(mol)
+    np.testing.assert_equal(charges, cached_charges)
+
+    new_charges, vjp_fn = jax.vjp(functools.partial(am1h.partial_parameterize, None, mol))
+
+    # charges_adjoints = np.random.randn(*charges.shape)
```
Comment (proteneer) [inline tests/test_handlers.py:482]: stray comment

File: tests/test_handlers.py [hunk 12]
```diff
@@ -456,6 +456,34 @@ def test_am1bcc_parameterization():
...
+def test_resp_parameterization():
+    # currently takes no parameters
+    smirks = []
+    params = []
+    props = None
+
+    cache_key = nonbonded.RESP
+    am1h = nonbonded.RESPHandler(smirks, params, props)
+    mol = Chem.AddHs(Chem.MolFromSmiles("C1CNCOC1F"))
+    AllChem.EmbedMolecule(mol)
```
Comment (proteneer) [inline tests/test_handlers.py:468]: If RESP to be independent of the initial conformer, we should explicitly be calling `mol.RemoveAllConformers()` beforehand as opposed to `EmbedMolecule()`

File: tests/test_handlers.py [hunk 13]
```diff
@@ -456,6 +456,34 @@ def test_am1bcc_parameterization():
     # assert vjp_fn(charges_adjoints) == None
 
 
+def test_resp_parameterization():
```
Comment (proteneer) [inline tests/test_handlers.py:459]: We should add a test for determinism (either in the presence or absence of a seed) since the underlying algorithm uses a stochastic conformer generator. i.e. we want to make sure that calling RESP multiple items returns identical charges on *fresh* non-cached molecules.

File: timemachine/ff/handlers/nonbonded.py [hunk 14]
```diff
@@ -1,27 +1,44 @@
...
 import networkx as nx
 import numpy as np
+from Auto3D.ASE.geometry import opt_geometry
 from jax import jit, vmap
 from numpy.typing import NDArray
+from openff.recharge.aromaticity import AromaticityModel
+from openff.recharge.utilities.molecule import extract_conformers
+from openff.toolkit import unit, RDKitToolkitWrapper
+from openff.toolkit.topology import Molecule
+from openff.toolkit.utils import AntechamberNotFoundError
```
Comment (proteneer) [inline timemachine/ff/handlers/nonbonded.py:21]: We probably don't have to import this

File: timemachine/ff/handlers/nonbonded.py [hunk 15]
```diff
@@ -1,27 +1,44 @@
...
+from openff.units import Quantity
```
Comment (proteneer) [inline timemachine/ff/handlers/nonbonded.py:22]: again, I think this is just what's in OpenMM

**Review by proteneer — COMMENTED** [review]
Two major things to discuss offline

1) openff toolkit/recharge dependency - I think this is mostly from _elf generation.
2) Auto3D/AIMNET dependency (which adds torch, torchvision, torchaudio).

File: timemachine/ff/handlers/nonbonded.py [hunk 16]
```diff
@@ -150,6 +169,297 @@ def oe_assign_charges(mol, charge_model=AM1BCCELF10):
...
+        mol = gto.Mole()
+        mol.cart = True  # Use Cartesian basis functions
+        mol.charge = Chem.GetFormalCharge(rdmol)
+        mol.spin = NumRadicalElectrons(rdmol)  # Number of unpaired electrons
+        mol.atom = xyz  # Atomic coordinates
+        mol.basis = "6-31gs"  # Standard basis set for RESP calculations
+        mol.build()
+
+        # Perform Restricted Hartree-Fock calculation on GPU
+        mf = scf.RHF(mol).to_gpu()
```
Comment (proteneer) [inline timemachine/ff/handlers/nonbonded.py:314]: Is there a meaningful speed difference of cpu vs gpu? We've never tested downstream timemachine code robustly in multi-CUDA context settings, and generally assumes that tm is the only library within the process that creates/initializes the underlying cuda context.
